src/directives/offense/poisonRoom.ts

```diff
@@ -0,0 +1,186 @@
+	created: number;
+	safeSince: number;
+}
+
+/**
+ * Poison sources in remote rooms by claiming controller, walling in its sources and controller then unclaiming it.
+ */
+@profile
+export class DirectivePoisonRoom extends Directive {
+	
+	static directiveName = 'PoisonRoom';
+	static color = COLOR_RED;
+	static secondaryColor = COLOR_BROWN;
+	static requiredRCL = 4;
+	
+	memory: DirectivePoisonRoomMemory;
+
+	walkableSourcePosisions: RoomPosition[];
+	walkableControllerPosisions: RoomPosition[];
+
+	overlords: {
+		roomPoisoner: RoomPoisonerOverlord;
+		scout: StationaryScoutOverlord;
+		defenders: OutpostDefenseOverlord;
+	};
+
+	constructor(flag: Flag) {
+		super(flag, colony => colony.level >= DirectivePoisonRoom.requiredRCL);
+
+		this.memory.poisonSourcesOnly = false;
+		this.memory.removeIfPoisoned = true;
+		this.memory.isPoisoned = this.isPoisoned();
+		this.memory.isHostile = (this.room && this.room.hostiles.length > 0)? true:false;
+
+		// Remove if misplaced
+		if (Cartographer.roomType(this.pos.roomName) != ROOMTYPE_CONTROLLER) {
+			log.warning(`${this.print}: ${printRoomName(this.pos.roomName)} is not a controller room; ` +
+						`removing directive!`);
+			this.remove(true);
+		}
+
+		//remove if created in owned room! (fail safe)
+		if(this.room && this.room.my && this.room.controller!.level > 2){
+			log.warning(`${this.print}: ${printRoomName(this.pos.roomName)} can not contaminate owned room; ` +
+			`manually unclaim it  first if you still want to contaminate it!`);
+			this.remove(true);
+		}
+		
+		// remove if already contaminated (if visible)
+		if (this.memory.removeIfPoisoned && this.isPoisoned()) {
+			log.warning(`${this.print}: ${printRoomName(this.pos.roomName)} is already contaminated; ` +
+						`removing directive!`);
+			this.remove(true);
+		}
+		
+	}
+
+	spawnMoarOverlords() {
+		//if room is visible + not claimed + GCL not enough, do not spawn roomPoisiner
+		if(this.pos.isVisible && this.room && !this.room.my && _.values(Overmind.colonies).length >= Game.gcl.level){
+			log.warning(`${this.print}: ${printRoomName(this.pos.roomName)} not enough GCL to contaminate room;`);
+			return;
+		}
+		if((this.room && this.room.hostiles.length > 0) || this.memory.isHostile){
+			this.memory.isHostile = true; //once hostile, always hostile, do not send scouts and always send defenders
+			this.overlords.defenders = new OutpostDefenseOverlord(this);
+		} else{
+			this.overlords.scout = new StationaryScoutOverlord(this);
+		}
+		this.overlords.roomPoisoner = new RoomPoisonerOverlord(this);
+	}
+
+	init() {
+		this.alert(`Poisoning Room ${this.pos.roomName}`);
+		//calculate wall positions
+		if(this.room && this.room.controller){
+			this.walkableSourcePosisions = _.filter(_.flatten(_.map(this.room.sources, s => s.pos.neighbors)),pos => pos.isWalkable(true));
+			this.walkableControllerPosisions =  _.filter(this.room.controller!.pos.neighbors, pos => pos.isWalkable(true));
+		}
+	}
+
+	private poison() {
```

Comment:
any actions related to spawning or controlling creeps should always be put on an overlord, not a directive. 

any actions related to removing or suspending the entire directive should be put an a directive, not an overlord.

Reply:
as a general rule, if you are accessing `this.overlords.X.property` on a directive, whatever is doing that should be on the overlord

Reply:
go it, will do it an update you